source/fuzz/transformation_add_opphi_synonym.cpp

```diff
@@ -0,0 +1,179 @@
+      fuzzerutil::MapToRepeatedUInt32Pair(preds_to_ids);
+  message_.set_fresh_id(fresh_id);
+}
+
+bool TransformationAddOpPhiSynonym::IsApplicable(
+    opt::IRContext* ir_context,
+    const TransformationContext& transformation_context) const {
+  // Check that |message_.block_id| is a block label id.
+  auto block = fuzzerutil::MaybeFindBlock(ir_context, message_.block_id());
+  if (!block) {
+    return false;
+  }
+
+  // Check that |message_.fresh_id| is actually fresh.
+  if (!fuzzerutil::IsFreshId(ir_context, message_.fresh_id())) {
+    return false;
+  }
+
+  // Check that |message_.pred_to_id| contains a mapping for all of the block's
+  // predecessors.
+  std::vector<uint32_t> predecessors = ir_context->cfg()->preds(block->id());
+
+  // There must be at least one predecessor.
+  if (predecessors.empty()) {
+    return false;
+  }
+
+  std::map<uint32_t, uint32_t> preds_to_ids =
+      fuzzerutil::RepeatedUInt32PairToMap(message_.pred_to_id());
+
+  // There must not be repeated key values in |message_.pred_to_id|.
+  if (preds_to_ids.size() != static_cast<size_t>(message_.pred_to_id_size())) {
+    return false;
+  }
+
+  // Check that each predecessor has a corresponding mapping.
+  for (uint32_t pred : predecessors) {
+    if (preds_to_ids.count(pred) == 0) {
+      return false;
+    }
+  }
+
+  // Check that all the ids exist in the module.
+  for (auto& pair : message_.pred_to_id()) {
+    if (!ir_context->get_def_use_mgr()->GetDef(pair.second())) {
+      return false;
+    }
+  }
+
+  // Check that all the ids are synonymous, they all have the same type and they
+  // are available to use at the end of the corresponding predecessor block.
+  uint32_t first_id = preds_to_ids.begin()->second;
+  uint32_t type_id = ir_context->get_def_use_mgr()->GetDef(first_id)->type_id();
+
+  // Check that the type of the id is allowed.
+  if (!CheckTypeIsAllowed(ir_context, type_id)) {
+    return false;
+  }
+
+  for (auto& pair : preds_to_ids) {
+    // Check that the id is synonymous with the others by checking that it is
+    // synonymous with the first one.
+    if (!transformation_context.GetFactManager()->IsSynonymous(
+            MakeDataDescriptor(pair.second, {}),
+            MakeDataDescriptor(first_id, {}))) {
+      return false;
+    }
+
+    // Check that the id has the same type as the other ones.
+    if (ir_context->get_def_use_mgr()->GetDef(pair.second)->type_id() !=
+        type_id) {
+      return false;
+    }
+
+    // Check that the id is available at the end of the corresponding
+    // predecessor block.
+
+    auto pred_block = ir_context->get_instr_block(pair.first);
+    // We should always be able to find the predecessor block, since it is in
+    // the predecessors list of |block|.
+    assert(pred_block && "Could not find one of the predecessor blocks.");
+
+    if (!fuzzerutil::IdIsAvailableBeforeInstruction(
+            ir_context, pred_block->terminator(), pair.second)) {
+      return false;
+    }
```

Comment:
This is a common issue. This function always returns false if `pred_block` is unreachable.

Reply:
Please, add a TODO(#3722) comment explaining that we need to refactor this. Use an original issue URL instead of #...

Reply:
Done